polkadot/xcm/xcm-executor/integration-tests/src/lib.rs

```diff
@@ -323,3 +325,84 @@ fn query_response_elicits_handler() {
 		)));
 	});
 }
+
+/// Simulates a cross-chain message from Parachain to Parachain through Relay Chain
+/// that deposits assets into the reserve of the destination.
+/// Regression test for `DepostiReserveAsset` changes in
+/// <https://github.com/paritytech/polkadot-sdk/pull/3340>
+#[test]
+fn deposit_reserve_asset_works_for_any_xcm_sender() {
```

Comment:
The previous `PriceForMessageDelivery` was flaky returning different prices.

Primarily, it was designed to return the same price for identical messages. However, reanchored messages weren't actually the same.
  - First price calculation is done for a message with the full asset amount in `ReserveAssetDeposited`
  - Actual price calculation during `send()` is done for a message where `transport_fee` has been deducted from the initial asset amount.
  
It does not explain by itself why the test was flaky. The reason for this flakiness is that the appended `SetTopic(id)` XCM instruction is calculated using the previous block as entropy. This is why any change in the runtime led to changes in the final message which might or might not generate enough transport fees to be paid.

All of this made me realise that the current method of setting aside the `transport_fee` for `DepositReserveAsset` will never be infallible for any possible `PriceForMessageDelivery` implementation. There is a circular dependency where it is impossible to know the actual `DepositReserveAsset` amount that will finally be included in the message from which the `transport_fee` is calculated. @acatangiu @franciscoaguirre 

I followed a different approach to fix the test.

Reply:
Can you expand a bit on the transport cost variance?

The actual amount (scalar value) of the asset within the message changes/impacts the message weight? I.e. An instruction containing asset: 2 DOT is more expensive than same instruction with asset: 3 DOT? Did I understand that correctly?
Or is it an issue with `SetTopic(id)` whose transport cost is not correctly calculated during execution of `ReserveAssetDeposited`? Looking at the code I'm missing where this cost changes between when we calculate and when it is charged..

Reply:
> An instruction containing asset: 2 DOT is more expensive than same instruction with asset: 3 DOT? Did I understand that correctly?

Not for our `PriceForMessageDelivery` implementation, because it only depends on the message size, that is why it was failing when the asset was not reanchored. 

I am just saying that potentially, a third party could make use of the `xcm-executor` with a different `PriceForMessageDelivery` implementation, and we can not guarantee that the set aside `transport_fee` approach for `DepositReserveAsset` will be infalible. It is just something to be aware of in case the design can be improved, not a big deal, not even probable, and something that shouldn't block this PR.

Reply:
I'll proceed to merge it if the new test looks ok for you.